src/lib/mongoBackend/mongoRegistrationCreate.h

```diff
@@ -0,0 +1,50 @@
+#ifndef SRC_LIB_MONGOBACKEND_MONGOREGISTRATIONCREATE_H_
+#define SRC_LIB_MONGOBACKEND_MONGOREGISTRATIONCREATE_H_
+
+/*
+*
+* Copyright 2017 Telefonica Investigacion y Desarrollo, S.A.U
+*
+* This file is part of Orion Context Broker.
+*
+* Orion Context Broker is free software: you can redistribute it and/or
+* modify it under the terms of the GNU Affero General Public License as
+* published by the Free Software Foundation, either version 3 of the
+* License, or (at your option) any later version.
+*
+* Orion Context Broker is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Ken Zangelin
+*/
+#include <string>
+#include <vector>
+
+#include "rest/HttpStatusCode.h"
+#include "rest/OrionError.h"
+#include "apiTypesV2/Registration.h"
+
+
+
+/* ****************************************************************************
+*
+* mongoRegistrationCreate - 
+*/
+extern void mongoRegistrationCreate
+(
+  ngsiv2::Registration*  regP,
+  const std::string&     tenant,
+  const std::string&     servicePathV,
+  std::string*           regIdP,
+  OrionError*            oeP
+);
+
```

Comment:
Looking to subscription creation function:

```
extern std::string mongoCreateSubscription
(
  const ngsiv2::Subscription&      sub,
  OrionError*                      oe,
  const std::string&               tenant,
  const std::vector<std::string>&  servicePathV,
  const std::string&               xauthToken,
  const std::string&               fiwareCorrelator
);
```

I understand we don't need correlator and token this time, as a consecuence of a registration no further message can be sent (different to subscription, that could have "initial notifications" synchronous to subscription creation).

With regards to servicepath, curiously subscription create uses a vector... although internally it seems the first element is used and the other ignored.

```
  std::string        servicePath      = servicePathV[0] == "" ? DEFAULT_SERVICE_PATH_QUERIES : servicePathV[0];
```

How this works in the case of registration? Have you getting that first element in a previous stage (e.g. rest layer or parsing)? The filal `V` in the variable name should be removed anycase.

Reply:
It seems I have found it myself :)

https://github.com/apinf/fiware-orion/pull/14/files#diff-0917c685f394e386632bfbd6640d1e65R85

If this is the strategy we want, then do the other way around. I mean, modify mongoCreateSubscription.cpp and postSubscriptions.cpp to use that way. It should be pretty easy, so I'd suggest not delaying and do it in this very PR.

Reply:
You mean to send not the entire vector to mongoCreateSubscription etc, but just the first element, right?

Reply:
That's a good point... how many calls to mongoCreateSubscription are (and from which place) in the code?

Reply:
I looked it up yesterday ...   It's called from 3-4 places (not including unit tests)

Reply:
Actually, 2 places:

From `postSubscription()`:

```
  TIMED_MONGO(subsID = mongoCreateSubscription(
                          parseDataP->subsV2,
                          &beError,
                          ciP->tenant,
                          ciP->servicePathV,
                          ciP->httpHeaders.xauthToken,
                          ciP->httpHeaders.correlator));
```

From `mongoSubscribeContext()` (as if you remember we refactor NGSIv1 mongo backend logic to be based on NGSIv2 logic):

```
std::string subId = mongoCreateSubscription(sub, &oe, tenant, servicePathV, xauthToken, fiwareCorrelator);
```

Thus I think it worth the change (and note that the same would probably happend with registrations, I mean, current NGSIv1 registration creation logic could be refactored in the future to be based on NGSIv2 so simplifying our code base).

Reply:
Looking into this a little ...
`mongoCreateSubscription` calls (with service path vector) `setCondsAndInitialNotify()`, that calls (with service path vector) `processConditionVector()`, etc.

This PR is already too big.

I propose to postpone this modification

Reply:
NTC, but an issue should be crated if we consider this of any importance.

Sending the entire vector or only the first item of the vector, both ways are sending only a pointer, so from the point of view of tput, it doesn't matter.
Making the check outside of the function forces us to duplicate code.
Making the check inside makes the function prototype strange, as it wants only a string but accepts a vector.
To be, avoiding to copy code is slightly more interesting than to have a 100% perfect function prototype, so I would keep it the way it is now. I have not a strong opinion here though, so if you want the change in the future, create the issue about it

Reply:
I can live with a 2-times duplication (if we need to duplicate 10 times it would be different). Thus, I chose the option of keep the propotype simpler and do it outsie, in the calling servicer routine.

However, no need to do the change in this PR is you prefer so. If this is to be done in the very next PR there isn't need of creating an issue.

Reply:
> If this is to be done in the very next PR there isn't need of creating an issue.

After looking close to this, it is no so easy...

The service path vector is passed down in the create subscription logic and ends in a entitiesQuery() call, that needs an actual vector (this function is also used by the entity query logic, where the vector has full sense). A large refactor would be needed and probably doesn't worth the effort by the time being.